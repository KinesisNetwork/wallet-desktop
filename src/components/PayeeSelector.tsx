import * as React from 'react'

import { changeTransferView as action } from '@actions'
import { FormChangeHandler, Payee, TransferRequest, TransferView } from '@types'

export interface Props {
  targetPayee: TransferRequest['targetPayee']
  payees: Payee[]
  handleChange: FormChangeHandler<TransferRequest>
  changeTransferView: typeof action
}

export const PayeeSelector: React.SFC<Props> = ({
  targetPayee,
  payees,
  handleChange,
  changeTransferView,
}) => {
  const isSelected = payees.findIndex(({ publicKey }) => targetPayee === publicKey) !== -1
  const selector =
    payees.length === 0 ? (
      <div className="control is-expanded">
        <a
          className="button is-fullwidth"
          onClick={() => changeTransferView(TransferView.addPayee)}
        >
          <span className="icon">
            <i className="fa fa-plus" />
          </span>
          <span>Add Payee</span>
        </a>
      </div>
    ) : (
      <React.Fragment>
        <div className="control is-expanded has-icons-left">
          <div className="select is-fullwidth">
            <select
              className={!isSelected ? 'has-text-grey' : ''}
              onChange={({ target: { value: newValue } }) =>
                handleChange({ field: 'targetPayee', newValue })
              }
              value={targetPayee}
            >
              <option value="" hidden={true}>
                Select a payee
              </option>
              <option value="">None</option>
              {payees.map(({ name, publicKey }) => (
                <option key={publicKey} value={publicKey}>
                  {name}
                </option>
              ))}
            </select>
          </div>
          <div className={`icon is-small is-left ${isSelected ? 'has-text-primary' : ''}`}>
            <i className="fas fa-user" />
          </div>
        </div>
        <div className="control">
          <a className="button" onClick={() => changeTransferView(TransferView.addPayee)}>
            <span className="icon">
              <i className="fa fa-plus" />
            </span>
          </a>
        </div>
      </React.Fragment>
    )
  return (
    <div className="field">
      <div className="field is-grouped">{selector}</div>
    </div>
  )
}

PayeeSelector.displayName = 'PayeeSelector'
