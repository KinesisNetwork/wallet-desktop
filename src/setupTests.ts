import { configure } from 'enzyme'
import * as Adapter from 'enzyme-adapter-react-16'

export default configure({ adapter: new Adapter() })
