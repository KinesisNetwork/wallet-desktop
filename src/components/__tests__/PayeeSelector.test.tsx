import { shallow } from 'enzyme';
import * as React from 'react'

import { PayeeSelector } from '@components/PayeeSelector'
import '../../setupTests'

describe('PayeeSelector', () => {
  let props

  beforeEach(() => {
    props = {
      payees: [],
      targetPayee: '',
      handleChange: () => null,
      changeTransferView: () => null
    }
  })

  it('should render correctly', () => {
    const wrapper = shallow(<PayeeSelector {...props} />)

    expect(wrapper).toMatchSnapshot()
  })

  describe('if no payess exist', () => {
    it('renders a button', () => {
      const wrapper = shallow(<PayeeSelector {...props} />)

      const button = wrapper.find('a')

      expect(button.hasClass('button')).toBe(true)
      expect(button.children().find('.icon')).toHaveLength(1)
      expect(button.text()).toEqual('Add Payee')
    })

    it('calls changeTransferView on clicking the button', () => {
      const changeTransferViewMock = jest.fn()
      const wrapper = shallow(<PayeeSelector {...props} changeTransferView={changeTransferViewMock} />)

      const button = wrapper.find('a')
      button.simulate('click')

      expect(changeTransferViewMock).toHaveBeenCalledWith(1)
    })
  })

  describe('if payees exist', () => {
    let newPayee

    beforeEach(() => {
      newPayee = [
        {
          name: 'My account',
          publicKey: '123'
        }
      ]
    })

    it('returns a select dropdown and a button', () => {
      const wrapper = shallow(<PayeeSelector {...props} payees={newPayee} />)

      expect(wrapper.find('select')).toHaveLength(1)
      expect(wrapper.find('.button')).toHaveLength(1)
    })

    it('renders N + 2 option fields, where N = number of payees', () => {
      const wrapper = shallow(<PayeeSelector {...props} payees={newPayee} />)

      const option = wrapper.find('option')

      expect(option).toHaveLength(3)
      expect(option.at(0).text()).toEqual('Select a payee')
      expect(option.at(0).prop('hidden')).toEqual(true)
      expect(option.at(1).text()).toEqual('None')
      expect(option.at(2).text()).toEqual('My account')
      expect(option.at(2).prop('value')).toEqual('123')
    })

    it('greyes out select field if no payees are selected', () => {
      const wrapper = shallow(<PayeeSelector {...props} payees={newPayee} />)

      expect(wrapper.find('select').hasClass('has-text-grey')).toBe(true)
    })

    it('calls the handleChange method on changing option', () => {
      const handleChangeMock = jest.fn()
      const handleChangeMockParam = {
        field: 'targetPayee',
        newValue: '123'
      }
      const wrapper = shallow(<PayeeSelector {...props} payees={newPayee} handleChange={handleChangeMock} />)

      const select = wrapper.find('select')
      select.simulate('change', { target: { value: '123' } })

      expect(handleChangeMock).toHaveBeenCalledWith(handleChangeMockParam)
    })
  })
})
