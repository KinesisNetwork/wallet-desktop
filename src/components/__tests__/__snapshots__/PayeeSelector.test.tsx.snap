// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PayeeSelector should render correctly 1`] = `
<div
  className="field"
>
  <div
    className="field is-grouped"
  >
    <div
      className="control is-expanded"
    >
      <a
        className="button is-fullwidth"
        onClick={[Function]}
      >
        <span
          className="icon"
        >
          <i
            className="fa fa-plus"
          />
        </span>
        <span>
          Add Payee
        </span>
      </a>
    </div>
  </div>
</div>
`;
